packages/contracts/src/contracts/current/forwarder/Forwarder.sol

```diff
@@ -0,0 +1,315 @@
+        if (totalFeeAmount > 0) {
+            // Fees are required for these orders. Buy enough ZRX to cover the future fill
+            Exchange.FillResults memory feeTokensResult = buyFeeTokens(feeOrders, feeSignatures, totalFeeAmount);
+            totalTakerAmountSpent = safeAdd(totalTakerAmountSpent, feeTokensResult.takerAssetFilledAmount);
+        }
+        address makerTokenAddress;
+        uint256 tokenId;
+        for (uint256 n = 0; n < orders.length; n++) {
+            Exchange.FillResults memory fillOrderResults = EXCHANGE.fillOrder(
+                orders[n],
+                orders[n].takerAssetAmount,
+                signatures[n]
+            );
+            // Fail if it wasn't fully filled otherwise we will keep WETH
+            require(fillOrderResults.takerAssetFilledAmount == orders[n].takerAssetAmount, ERROR_FAILED_TO_FILL_ALL_ORDERS);
+            totalTakerAmountSpent = safeAdd(totalTakerAmountSpent, fillOrderResults.takerAssetFilledAmount);
+            makerTokenAddress = readAddress(orders[n].makerAssetData, 1);
+            tokenId = readUint256(orders[n].makerAssetData, 21);
+            transferNFTToken(makerTokenAddress, msg.sender, tokenId);
+        }
+        // Prevent a user from over estimating the amount of ETH required
+        require(isAcceptableThreshold(msg.value, totalTakerAmountSpent), ERROR_UNACCEPTABLE_THRESHOLD);
+    }
+
+    function marketSellTokens(
+        Order[] memory orders,
+        bytes[] memory signatures,
+        Order[] memory feeOrders,
+        bytes[] memory feeSignatures,
+        uint256 sellTokenAmount)
+        private
+        returns (Exchange.FillResults memory totalFillResult)
+    {
+        uint256 takerTokenBalance = sellTokenAmount;
+        address makerTokenAddress = readAddress(orders[0].makerAssetData, 1);
+
+        Exchange.FillResults memory tokensSellQuote = marketSellOrdersQuote(orders, sellTokenAmount);
+        Exchange.FillResults memory requestedTokensResult;
+        if (tokensSellQuote.takerFeePaid > 0) {
+            // Fees are required for these orders
+            // Buy enough ZRX to cover the future market sell
+            Exchange.FillResults memory feeTokensResult = buyFeeTokens(feeOrders, feeSignatures, tokensSellQuote.takerFeePaid);
+            takerTokenBalance = safeSub(takerTokenBalance, feeTokensResult.takerAssetFilledAmount);
+            // Make our market sell to buy the requested tokens with the remaining balance
+            requestedTokensResult = EXCHANGE.marketSellOrders(orders, takerTokenBalance, signatures);
+            // It's possibile to over-buy fees by a small amount as the marketSellQuote was based on 100% of ETH
+            // and marketSell (after fee abstraction) less than 100%. If a user uses this enough it may be worth withdrawing
+            balanceOf[msg.sender] += safeSub(feeTokensResult.makerAssetFilledAmount, requestedTokensResult.takerFeePaid);
+            require(balanceOf[msg.sender] >= 0);
```

Comment:
When could this ever by negative?

Reply:
Also, please add error message.

Reply:
Unlikely to happen but it is output from another contract so it's best to check that.